Extend documentation for smart contract queries (#659)
Concordium/concordium.github.io: source/mainnet/smart-contracts/general/develop-contracts.rst

Working with queries
--------------------

Queries can be called from smart contracts to query an account balance, contract balance,
or the current exchange rates.

Query an account balance
~~~~~~~~~~~~~~~~~~~~~~~~

To query account balances, the following are available:

.. code-block:: rust

   // Query the balance of an account.
   let account_balance = host.account_balance(account_address)?;

Assuming the account exists, this will return all of the public balances of an account,
which can then be accessed as:

.. code-block:: rust

   // The amount which is available for transfers.
   let available_balance = account_balance.available();
   // The staked amount.
   let staked_balance = account_balance.staked();
   // The amount locked in scheduled transfers.
   let locked_balance = account_balance.locked();
   // The total public balance, i.e. including staked and locked_balance.
   let total_balance = account_balance.total();

.. note::

   When sending a smart contract update transaction, the invoker provides a max energy cost for the execution.
   CCD equivalent to the max energy cost is then deposited during execution of the contract.

>    Because of this, querying the balance of the invoker will result in the current account balance subtracted

Reviewer: ```suggestion
   Because of this, querying the balance of the invoker will result in the current account balance minus
```

>    amount of CCD needed to cover the max energy cost.

Reviewer 2: We should also document the semantics of this query with respect to transfers/invoke calls inside a single transaction.

Reviewer 2: That is alluded to for contracts, but not here.

Author: I have added to this

Reviewer 2: Thanks.

>    with the amount of CCD needed to cover the max energy cost.

Reviewer: ```suggestion
   the amount of CCD needed to cover the max energy cost.
```

Query a contract balance
~~~~~~~~~~~~~~~~~~~~~~~~

To query the current balance of a contract, the following is available:

.. code-block:: rust

    let contract_balance = host.contract_balance(address)?;

> Assuming the contract exists, this will return the current amount hold by the contract.

Reviewer: ```suggestion
Assuming the contract exists, this will return the current amount held by the contract.
```

> Any amount transferred or received during the same transaction uptil the point of querying

Reviewer: ```suggestion
Any amount transferred or received during the same transaction until the point of querying
```

> are be reflected in the balance.

Reviewer: ```suggestion
is reflected in the balance.
```

.. note::

>    Although it is valid for a contract to query its own balance, it cheaper to use ``host.self_balance()``.

Reviewer 3: ```suggestion
   Although it is valid for a contract to query its own balance, it is cheaper to use ``host.self_balance()``.
```

   let locked_balance = account_balance.locked();
   // The total public balance, i.e. including staked and locked_balance.
   let total_balance = account_balance.total();

Query a contract balance
~~~~~~~~~~~~~~~~~~~~~~~~

To query the current balance of a contract, the following is available:

.. code-block:: rust

    let contract_balance = host.contract_balance(address)?;

Assuming the contract exists, this returns the current amount held by the contract.
Any amount transferred or received during the same transaction until the point of querying
is reflected in the balance.

.. note::

   Although it is valid for a contract to query its own balance, it is cheaper to use ``host.self_balance()``.

Query exchange rates
~~~~~~~~~~~~~~~~~~~~

To query exchange rates, the following are available:

.. code-block:: rust

    let exchange_rates = host.exchange_rates();

The result contains the exchange rates currently used by the chain.
These are the micro CCD per Euro rate and the Euro per NRG rate.

.. code-block:: rust

    let micro_ccd_per_euro = exchange_rates.micro_ccd_per_euro();
    let euro_per_energy = exchange_rates.euro_per_energy();

> Each rate is a ratio, with a numerator and a denominator.

Reviewer 2: ```suggestion
Each rate is a ratio of two 64-bit unsigned integers.
```
